R/restore.R

```diff
@@ -65,6 +71,39 @@ drop_missing_deps <- function(deps) {
   lapply(deps, intersect, pkgs)
 }
 
+#' Drop packages that do not need an update
+#'
+#' If any of the following are true, then a package is needed; otherwise it
+#' unneeded. The package is not installed, or the currently installed version of
+#' the package is not the same as the target version, or the currently installed
+#' package was not built with the current version of R.
+#'
+#' @param pkgs A data frame with columns Package and Version
+#' @return A (filtered) data frame with columns Package and Version
+#'
+#' @keywords internal
+
+drop_unneeded_pkgs <- function(pkgs) {
+  cur_pkgs <- as.data.frame(installed.packages(), stringsAsFactors = FALSE,
+                            row.names = FALSE)
+  cur_pkgs$Built <- as.package_version(cur_pkgs$Built)
+
+  is_needed <- function(package, version) {
+    if (!(package %in% cur_pkgs$Package) ||
+        cur_pkgs$Version[cur_pkgs$Package == package] != version ||
+        cur_pkgs$Built[cur_pkgs$Package == package] != getRversion())
```

Comment:
If we have multiple libraries, with different versions of the same package, then this condition might not be correct. (I assume that `installed.packages()` lists all versions from all libraries, but I haven't tried.)

Maybe we should indeed restrict this to a single library?